### ISSUE TITLE
DiffusionGemma denoise attends outside the HF sliding window on 25/30 layers (P>=1024)

### ISSUE THREAD

**Issue by zhenghaoniTT** [tenstorrent/tt-metal#51080]
## Summary

DiffusionGemma's TT denoise path lets **all 30 decoder layers attend the entire committed KV prefix**, but in HF DiffusionGemma the 25 sliding layers only ever hold the last `sliding_window - 1 = 1023` committed tokens. So for any prompt past 1024 tokens, TT attends keys that the HF reference **does not have in its cache at all**, on 25 of 30 layers.

This is present in the shipped serving configuration (`doc/optimize_perf/run_upfront_gpqa.sh`, `--max-model-len 4096`, thinking prompts around 3072 tokens), i.e. well past the 1024 threshold.

Filing this as a correctness issue rather than a performance one, even though the fix is also ~2.4x cheaper at `p_max=4096` and ~5.7x at 128K, because the decision-changing part is a fidelity repair, not a tuning choice.

Related: #47452 (DG issue tree), #48291 (decision-fidelity bar — the gate for the fix, see below).

## Model geometry (verified against the checkpoint)

`diffusiongemma-26B-A4B-it/config.json`: 30 layers, `sliding_window = 1024`, `full_attention` at indices `[5, 11, 17, 23, 29]` — **5 full-attention layers, 25 sliding layers**.

## Evidence: what TT does today

`models/experimental/diffusion_gemma/tt/denoise_forward.py:1163-1171` — the persistent reveal mask is built with a hardcoded layer type, for every layer:

```python
def _build_reveal_mask_device(self, prompt_len: int):
    return build_device_canvas_reveal_mask(
        self.tt_model.mesh_device,
        prompt_len=prompt_len,
        canvas_len=self._reveal_canvas_len,
        p_max=self._reveal_p_max,
        layer_type="full_attention",          # <-- unconditional, all 30 layers
        enforce_sliding_window=self._reveal_enforce_window,   # <-- False in production
    )
```

`_reveal_mask_provider` (`:1197-1198`) returns that single buffer for every `layer_idx`. So sliding layers see the same all-attend mask over the full `p_max` span as the full-attention layers.

## Evidence: what HF actually does

Read from `transformers/models/diffusion_gemma/modeling_diffusion_gemma.py` in the model venv.

`create_diffusion_decoder_attention_mask` (`:1302-1439`) builds the sliding mask by **slicing the padding mask**, not by a distance predicate:

```python
sliding_start_idx = valid_cache_tokens - config.sliding_window + 1   # non-compileable cache
sliding_end_idx   = valid_cache_tokens
sliding_mask = full_mask[..., sliding_start_idx:sliding_end_idx]
sliding_mask = torch.nn.functional.pad(sliding_mask, (0, canvas_length), value=True)
```

The window itself comes from **cache truncation**, not from the mask. `DynamicSlidingWindowLayer.update` (`transformers/cache_utils.py:245-247`) keeps only:

```python
self.keys   = full_key_states[:, :, -self.sliding_window + 1 :, :]
self.values = full_value_states[:, :, -self.sliding_window + 1 :, :]
```

DG's default cache is `DynamicCache(config=...)` (`modeling_diffusion_gemma.py:943`). `sliding_window=` is forwarded to the attention interface (`:467`) but both `eager_attention_forward` and the sdpa integration ignore it.

**Ground truth: HF sliding-layer denoise visibility = `[P - 1023, P)` all-attend, plus the full canvas. `keep = sliding_window - 1 = 1023`. The window binds iff `P >= 1024`.** There is no per-query staircase — every canvas row sees the same key set.

## Impact

For `P >= 1024`, on 25 of 30 layers, TT attends `P - 1023` extra committed keys per denoise step that HF does not retain. At the shipped 4096 config with a 3072-token prompt that is ~2049 extra keys per sliding layer per step, for 48 steps per block.

The decision-level impact is **not yet quantified** — that measurement is part of the fix's gate, not a precondition for filing. Note that #48291 established the bf16 decision-agreement floor separately; this is a distinct, structural divergence and should not be conflated with that precision analysis.

## Secondary bug: the repo's reference encodes a rule HF does not implement

`models/experimental/diffusion_gemma/reference/attention_mask.py:11-17` documents HF's behaviour as:

> ... so they attend only keys with `abs(q_idx - kv_idx) <= sliding_window`

and `:78-81` implements exactly that staircase:

```python
if enforce_sliding_window and layer_type == "sliding_attention":
    allowed = allowed & ((q_abs - k_abs).abs() <= sliding_window)
```

**HF does not apply this.** Enabling `enforce_sliding_window=True` as-is would hide up to 255 keys per canvas row that HF *does* attend — trading one fidelity bug for another. The same wrong rule is encoded in:

- `build_canvas_denoise_mask(layer_type="sliding_attention")`, `reference/attention_mask.py:158-163`
- `_sliding_layer_needs_denoise_mask`, `tt/denoise_forward.py:153-154` — its threshold `prompt_len + canvas_len - 1 > sliding_window` follows from the staircase; the correct threshold is simply `P >= 1024`.

Any prior A/B validated against these builders was validated against the wrong reference.

## Third observation: internal inconsistency with prefill

gemma4 prefill **does** enforce the window — `chunked_prefill_sdpa_sliding` passes `sliding_window_size=sliding_window` (`models/demos/gemma4/tt/attention/operations.py:373`). So the prompt KV is produced under windowed attention and then consumed unwindowed by denoise. That is inconsistent independent of what HF does.

## Proposed fix

Per-layer prefix spans:

- 25 sliding layers: read a constant 1024 rows at `lo = max(0, P - 1024)` (P is always 32-aligned, so `lo` is too and satisfies the existing tile-alignment guard in `read_prompt_kv_cache_slice`). Per-layer block-resident K/V buffers, refreshed **outside** the trace so the sliding offset never enters a captured shape.
- 5 full-attention layers: unchanged, keep `p_max`.
- Two reveal masks instead of one: `[1,1,256,p_max+256]` for full layers, `[1,1,256,1280]` for sliding layers with content `lo + r < P` AND `lo + r >= P - 1023`, canvas all-attend. In steady state (`P >= 1024`) the sliding mask is `P`-independent — it stops changing after the first block.
- Correct the reference docstring, the `enforce_sliding_window` branch, and `_sliding_layer_needs_denoise_mask` in the same change.

Rejected: ttnn SDPA's built-in `sliding_window_size`. For `!is_causal` it is a window *centred* on the query index in the K axis (`sliding_window_geometry.hpp`), and DG denoise is a 256-row Q axis against a `(span + 256)` K axis at an arbitrary offset, so the op anchors on the wrong keys.

## Gating

Two tiers, because the change is bit-exact in one regime and decision-changing in the other:

1. `P <= 1023` — window never binds. Require **bitwise** `committed_sha256` identical to the current golden.
2. `P >= 1024` — decision-changing. Measure decision agreement against **fp32 HF DiffusionGemma**, not against today's TT output, since today's TT output is the thing being corrected. Agreement is expected to *improve*.

Must not be bundled with any other change, so the decision delta is attributable.

Implementation risk to close explicitly: a stale block-resident prefix buffer would silently replay old prompt KV. Carry a fail-loud `(cache_generation, prompt_len)` stamp on the reader and assert it before every block.

**Comment by zhenghaoniTT** [comment]
## Root cause of the wrong reference rule

While implementing the fix I found how the `abs(q_abs - k_abs) <= sliding_window` staircase entered this repo, and it is worth recording because it is a reusable trap.

`tests/test_real_transformers_parity.py::test_canvas_denoise_mask_matches_real_bidirectional_layer_masks` validated the **decoder's denoise** mask against `transformers.masking_utils.create_masks_for_generate` — the **generic** transformers mask builder. DiffusionGemma's identically-named override lives on `DiffusionGemmaEncoderModel` (class opens at `modeling_diffusion_gemma.py:979`, method at `:1138`) and simply delegates to that generic function.

That function describes the **encoder's** prompt self-attention, where the sliding window genuinely *is* a per-(query, key) distance predicate — so its sliding mask is a staircase, and the test passed.

But the denoise pass is the **decoder**, which builds masks with `DiffusionGemmaDecoderModel.create_diffusion_decoder_attention_mask` (`:1301`). That applies no staircase at all: on the ordinary unpadded `DynamicCache` path it early-returns `{"full_attention": None, "sliding_attention": None}`, and the window is purely cache truncation (`DynamicSlidingWindowLayer.update` retains `-sliding_window + 1` rows). When a padding mask *is* materialized it is expanded from a 1-D per-key vector, so it has no query-index dependence.

**So the decoder was validated against the encoder's reference.** Both HF functions are correct for their own layer; picking the wrong one silently produced a plausible-looking mask with the wrong shape of dependence.

## Progress

Landed behind `DG_DENOISE_SLIDING_WINDOW` (default OFF, decision-changing above `P >= 1024`):

- reference corrected in `reference/attention_mask.py` (both builders) and `_sliding_layer_needs_denoise_mask`, whose threshold wrongly included `canvas_len` — an artifact of the staircase;
- per-layer-TYPE reveal masks in the denoise adapter, **all of one `[1,1,C,p_max+C]` shape**, so enabling the window changes mask content only and every captured trace stays shape-valid;
- HF semantics pinned CPU-only in `tests/test_hf_sliding_window_reference.py` (retention = `W-1`; no materialized mask on the unpadded path; no query dependence; TT span formula matches HF's key set);
- the parity test above now asserts the encoder staircase *as encoder behaviour* and points at the decoder pinning test.

The read span is still the full `p_max`, deliberately: that makes the fidelity fix need no new buffer machinery and no shape change. Bounding the sliding read to 1024 rows is a separate perf change, and it gets easier to gate afterwards — once out-of-window keys are already masked to `NEG`, not reading them changes only flash K-chunk accumulation order rather than decisions.

Remaining to flip the default: the decision-agreement run against **fp32 HF** (not against today's TT output, which is the defect being corrected). Design note: `doc/optimize_perf/per_layer_prefix_spans.md`.
